### PR TITLE
Issue 18887 set es custom mappings when a field is created

### PR DESCRIPTION
**Pull request by nollymar**
This change brings the benefit of setting an ES mapping on an index at the moment a field is created. Before, mapping were applied only when a full reindex was executed (specifically during index creation).

An ES mapping is set according to the field type and in case of failure, the save process should continue and the error ignored, just logged

Besides, some integration tests were implemented.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 1]
```diff
@@ -298,6 +301,61 @@ private void addMappingForFieldIfNeeded(final String indexName,
...
+     * @param field
+     * @param fieldVariableName
+     * @return
+     */
+    private String getMappingForField(final Field field, final String fieldVariableName) {
+        String mappingForField = null;
+        if (field instanceof DateField || field instanceof DateTimeField
+                || field instanceof TimeField) {
+            mappingForField = "{\n\"type\":\"date\",\n";
+            mappingForField += "\"format\": \"yyyy-MM-dd't'HH:mm:ss||MMM d, yyyy h:mm:ss a||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd||epoch_millis\"\n}";
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:315]: Codacy found an issue: [Prefer StringBuilder (non-synchronized) or StringBuffer (synchronized) over += for concatenating strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 2]
```diff
@@ -1357,6 +1362,129 @@ public void test_SaveField_GivenGraphQLCompatibleVariable_ShouldSave(
...
+
+            final IndiciesInfo indiciesInfo = APILocator.getIndiciesAPI().loadIndicies();
+            final ESMappingAPIImpl mappingAPI = new ESMappingAPIImpl();
+
+            //verify mapping on working index
+            Map<String, String> mapping = (Map<String, String>) mappingAPI
+                    .getFieldMappingAsMap(indiciesInfo.getWorking(),
+                            (type.variable() + StringPool.PERIOD + field.variable())
+                                    .toLowerCase()).get(field.variable());
+            assertTrue(UtilMethods.isSet(mapping.get("type")));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1434]: Codacy found an issue: [The String literal "type" appears 8 times in this file; the first occurrence is on line 1,434](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 3]
```diff
@@ -1357,6 +1362,129 @@ public void test_SaveField_GivenGraphQLCompatibleVariable_ShouldSave(
...
+                    .build();
+            field = fieldAPI.save(field, user);
+
+            final IndiciesInfo indiciesInfo = APILocator.getIndiciesAPI().loadIndicies();
+            final ESMappingAPIImpl mappingAPI = new ESMappingAPIImpl();
+
+            //verify mapping on working index
+            Map<String, Object> mapping = mappingAPI
+                    .getFieldMappingAsMap(indiciesInfo.getWorking(),
+                            (type.variable() + StringPool.PERIOD + field.variable()));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1392]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java [hunk 4]
```diff
@@ -77,6 +79,7 @@
 import com.liferay.portal.language.LanguageUtil;
 import com.liferay.portal.model.User;
 
+import java.io.IOException;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java:82]: Codacy found an issue: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

**Comment by erickgonzalez** [comment]
There is a test failing https://raw.githack.com/dotCMS/test-results/issue-18887-set-es-custom-mappings-when-a-field-is-created/projects/core/current/integration/postgres/reports/html/classes/com.dotmarketing.portlets.contentlet.business.ContentletAPITest.html#testUniqueTextFieldContentlets[1:%20com.dotmarketing.portlets.contentlet.business.ContentletAPITest$testCaseUniqueTextField@611c7b49] don't know if related to the changes but it's worth to double-check since it's not in the list of blinker failing tests.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 5]
```diff
@@ -92,6 +95,30 @@ public void addCustomMapping(final String indexName) {
...
+     * @param indexName - Index where mapping will be applied
+     * @param field
+     * @throws DotSecurityException
+     * @throws DotDataException
+     * @throws IOException
+     * @throws JSONException
+     */
+    @CloseDBIfOpened
+    public void addCustomMapping(final String indexName, Field field)
+            throws DotSecurityException, DotDataException, IOException, JSONException {
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:110]: param field could be marked final too.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 6]
```diff
@@ -298,6 +301,61 @@ private void addMappingForFieldIfNeeded(final String indexName,
...
+                mappingForField = "{\n\"type\":\"float\"\n}";
+            } else if (field.dataType() == DataTypes.INTEGER) {
+                mappingForField = "{\n\"type\":\"integer\"\n}";
+            } else if (!matchesExclusions(fieldVariableName)){
+                mappingForField = "{\n\"type\":\"text\"\n}";
+            }
+        }
+        return mappingForField;
+    }
+
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:331]: codacy already recommended using StringBuilder here and I like that but I think is more maintainable doing something like :

mappingForField =  String.format("{\n\"type\":\"%s\"\n}",fieldType);    

and pass fieldType. That way we can have that string format declared just once.

and the fieldType can be returned out of function so the function I imagine could use a static map and look like this:

```
ImmutableMap.of(
DataTypes.BOOL,DataTypes.BOOL.toString().toLowercase(),   //Could do something like this 
DataTypes.FLOAT, "float",   //or simply
DataTypes.INTEGER, "integer"
) ;
```
I've seen that Using maps like this to eliminate nested ifs. Is called the anti-if pattern.

**Review by fabrizzio-dotCMS — CHANGES_REQUESTED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 7]
```diff
@@ -1357,6 +1362,129 @@ public void test_SaveField_GivenGraphQLCompatibleVariable_ShouldSave(
...
+            ContentTypeDataGen.remove(type);
+        }
+    }
+
+    /**
+     * <b>Method to test:</b> {@link FieldAPIImpl#save(Field, User)}
+     * <p>
+     * <b>Given scenario:</b> A new {@link DateField} is saved and marked as `indexed=true`
+     * <p>
+     * <b>Expected result:</b> The field should be saved and mapped in ES with `type=date`
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1410]: maybe we can test other data types and their expected mapping

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java:82]: Codacy found an issue: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:315]: Codacy found an issue: [Prefer StringBuilder (non-synchronized) or StringBuffer (synchronized) over += for concatenating strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1392]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java:82]: Codacy found an issue: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1434]: Codacy found an issue: [The String literal "type" appears 8 times in this file; the first occurrence is on line 1,434](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1392]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 8]
```diff
@@ -298,6 +301,65 @@ private void addMappingForFieldIfNeeded(final String indexName,
...
+     * @param field
+     * @param fieldVariableName
+     * @return
+     */
+    private String getMappingForField(final Field field, final String fieldVariableName) {
+        String mappingForField = null;
+        if (field instanceof DateField || field instanceof DateTimeField
+                || field instanceof TimeField) {
+            mappingForField = "{\n\"type\":\"date\",\n";
+            mappingForField += "\"format\": \"yyyy-MM-dd't'HH:mm:ss||MMM d, yyyy h:mm:ss a||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd||epoch_millis\"\n}";
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:315]: Codacy found an issue: [Prefer StringBuilder (non-synchronized) or StringBuffer (synchronized) over += for concatenating strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

**Comment by erickgonzalez** [comment]
Seems that we have a few broken tests https://raw.githack.com/dotCMS/test-results/issue-18887-set-es-custom-mappings-when-a-field-is-created/projects/core/current/integration/postgres/reports/html/index.html

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 9]
```diff
@@ -299,6 +301,64 @@ private void addMappingForFieldIfNeeded(final String indexName,
...
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:315]: Codacy found an issue: [Prefer StringBuilder (non-synchronized) or StringBuffer (synchronized) over += for concatenating strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 10]
```diff
@@ -1390,6 +1395,129 @@ public void test_SaveField_GivenGraphQLCompatibleVariable_ShouldSave(
...
+                    .build();
+            field = fieldAPI.save(field, user);
+
+            final IndiciesInfo indiciesInfo = APILocator.getIndiciesAPI().loadIndicies();
+            final ESMappingAPIImpl mappingAPI = new ESMappingAPIImpl();
+
+            //verify mapping on working index
+            Map<String, Object> mapping = mappingAPI
+                    .getFieldMappingAsMap(indiciesInfo.getWorking(),
+                            (type.variable() + StringPool.PERIOD + field.variable()));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1425]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java:82]: Codacy found an issue: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java [hunk 11]
```diff
@@ -1390,6 +1395,129 @@ public void test_SaveField_GivenGraphQLCompatibleVariable_ShouldSave(
...
+
+            final IndiciesInfo indiciesInfo = APILocator.getIndiciesAPI().loadIndicies();
+            final ESMappingAPIImpl mappingAPI = new ESMappingAPIImpl();
+
+            //verify mapping on working index
+            Map<String, String> mapping = (Map<String, String>) mappingAPI
+                    .getFieldMappingAsMap(indiciesInfo.getWorking(),
+                            (type.variable() + StringPool.PERIOD + field.variable())
+                                    .toLowerCase()).get(field.variable());
+            assertTrue(UtilMethods.isSet(mapping.get("type")));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1467]: Codacy found an issue: [The String literal "type" appears 8 times in this file; the first occurrence is on line 1,434](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1425]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 11 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1467]: Codacy found an issue: [The String literal "type" appears 8 times in this file; the first occurrence is on line 1,434](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:315]: Codacy found an issue: [Prefer StringBuilder (non-synchronized) or StringBuffer (synchronized) over += for concatenating strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java:82]: Codacy found an issue: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 7 above)
Comment (nollymar) [reply dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1443]: those cases have been already tested

**Comment by nollymar** [comment]
> Seems that we have a few broken tests https://raw.githack.com/dotCMS/test-results/issue-18887-set-es-custom-mappings-when-a-field-is-created/projects/core/current/integration/postgres/reports/html/index.html

Broken tests have been fixed

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java (hunk 6 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:335]: Good suggestion. Done!

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1425]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java (hunk 11 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/business/FieldAPITest.java:1467]: Codacy found an issue: [The String literal "type" appears 8 times in this file; the first occurrence is on line 1,434](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/FieldAPIImpl.java:82]: Codacy found an issue: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 12]
```diff
@@ -299,6 +303,63 @@ private void addMappingForFieldIfNeeded(final String indexName,
...
+     */
+    private String getMappingForField(final Field field, final String fieldVariableName) {
+        final Map<DataTypes, String> dataTypesMap = ImmutableMap
+                .of(DataTypes.BOOL, "boolean", DataTypes.FLOAT, "float", DataTypes.INTEGER,
+                        "integer");
+        String mappingForField = null;
+        if (field instanceof DateField || field instanceof DateTimeField
+                || field instanceof TimeField) {
+            mappingForField = "{\n\"type\":\"date\",\n";
+            mappingForField += "\"format\": \"yyyy-MM-dd't'HH:mm:ss||MMM d, yyyy h:mm:ss a||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd||epoch_millis\"\n}";
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:320]: Codacy found an issue: [Prefer StringBuilder (non-synchronized) or StringBuffer (synchronized) over += for concatenating strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5866177)

**Review by fabrizzio-dotCMS — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java [hunk 13]
```diff
@@ -132,6 +155,25 @@ private void addCustomMappingForRelationships(final String indexName,
...
+     * @param relationshipName - Relationship to be indexed
+     * @throws JSONException
+     * @throws IOException
+     */
+    private void putRelationshipMapping(final String indexName, final String relationshipName)
+            throws JSONException, IOException {
+        final JSONObject properties = new JSONObject();
+        properties.put("properties", new JSONObject()
+                .put(relationshipName,
+                        new JSONObject("{\n"
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:170]: I wonder if we should not use something such as 
```
JSONObjectBuilder()
    .add("type","keyword")
   .add("ignore_above",8191).build()
```

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/ESMappingUtilHelper.java:170]: It seems the JSONObject has already methods to do it. Not need to create any builder

**Pull request merged**